### PR TITLE
Add '--weight' flag to 'promtool check metrics' command

### PR DESCRIPTION
**Pull request by chenlujjj**
Signed-off-by: chenlujjj <953546398@qq.com>

This PR is related to [issue](https://github.com/prometheus/prometheus/issues/10000). 

IMHO the flag fits better to be under `check metrics` subcommand compared to `debug metrics`. Since `debug metrics` just fetch metrics from a server and write them to a `debug.tar.gz` file. This `fetch-write` logic is coupled with `debug pprof` via `debugWrite` function.

An example:
image
image

### CODE REVIEW

**Comment by LeviHarrison** [comment]
Thanks for this!

I think the metrics should be in descending order because the ones the user will be interested in are the biggest.

**Review by JessicaGreben — CHANGES_REQUESTED** [review]
Thanks for the PR, it looks great. I like this idea a lot.  I left a few comments, also there are still lints errors, and it might also be a good idea to add a test or two.

File: cmd/promtool/main.go [hunk 1]
```diff
@@ -88,6 +92,7 @@ func main() {
 	).Required().ExistingFiles()
 
 	checkMetricsCmd := checkCmd.Command("metrics", checkMetricsUsage)
+	checkMetricsWeight := checkMetricsCmd.Flag("weight", "Check the weight of metrics.").Bool()
```
[the remark this answers is not in the document]
Comment (LeviHarrison) [reply cmd/promtool/main.go:95]: https://github.com/prometheus/prometheus/issues/10000#issuecomment-997289311

**Comment by chenlujjj** [comment]
Hi, @LeviHarrison , @JessicaGreben, I have updated the PR, including following update:
1. ranks metrics in descending order
2. add more info about the command usage
3. add unit test
4. fix lint issue

With respect to the command naming, I personally perfer the idea of @LeviHarrison since it does NOT focus on ONLY ONE metric.
And it seems like that I cannot create a subcommand of another subcommand.. I have tried to write like this:
`checkMetricsWeightCmd := checkMetricsCmd.Command("weight", checkMetricsWeightUsage)`
But in this way, the "promtool check metrics weight" and "promtool check metrics" command cannot both work. Not sure if I misused the kingpin library. Please point out if I made a mistake here. :)

**Comment by LeviHarrison** [comment]
> And it seems like that I cannot create a subcommand of another subcommand.

I think you're right, kingpin doesn't seem to support subcommands when the main command can be standalone. This leaves us with two options: `check metrics-weight` or `check metrics --weight`. I personally prefer the former, because I don't think a flag should change the function of a command that drastically. The hyphen might be weird though.

**Comment by LeviHarrison** [comment]
cc @beorn7 for feedback on the histogram and summary weight calculation

**Comment by JessicaGreben** [comment]
These changes look good to me. I like `check-weight` command like Levi suggested. 

I'm going to do another review, but also the tests aren't passing so that needs to be fixed:
Its failing saying :
```
        	Error Trace:	main_test.go:333
        	Error:      	Received unexpected error:
        	            	error while parsing text to metric families: text format parsing error in line 2: unknown metric type "summary\r"
        	Test:       	TestCheckMetricsWeight
--- FAIL: TestCheckMetricsWeight (0.00s)
```

**Comment by chenlujjj** [comment]
Thanks for the review comments. 
@JessicaGreben, I cannot reproduce the test failing in my local environment:
image

Is the failing message you posted related to the "ci/circleci: test_windows" CI step ?

**Comment by chenlujjj** [comment]
Oh, I see it. It seems to be caused by the different carriage return on MacOS and Windows. I will try to fix it.

**Comment by chenlujjj** [comment]
After a bit of investigation, I'm not sure if we should parse the metrics text on windows platform. The "prometheus/common" repository does not include tests on windows.
Another solution may be preparing two `metrics.prom` files, one for unix using LF as EOL, one for windows using CRLF as EOL. And the test case can open different file based on OS. But I think it is a little verbose here.

**Comment by brian-brazil** [comment]
Carriage returns are not valid line endings for Prometheus text format or
OpenMetrics, on any platform.

What you are calling weight is called cardinality in the Prometheus
ecosystem, it covers a count of anything.

**Comment by chenlujjj** [comment]
Then the test failure becomes a little confusing to me. I have confirmed that the `metrics-test.prom` file uses LF (line feed) as EOL.
image

**Comment by chenlujjj** [comment]
I think the `weight` here means the percentage of one metric's cardinality among all metrics cardinality sum.
Maybe the `count` field in `metricStat` struct should be renamed to `cardinality`. 
@LeviHarrison @JessicaGreben @brian-brazil What's your idea ?

**Comment by JessicaGreben** [comment]
This always seemed like cardinality to me, but @LeviHarrison disagreed so I will defer to what he thinks is best.

**Comment by LeviHarrison** [comment]
@brian-brazil in my opinion this command is less about the cardinality of individual metrics and more about how that cardinality contributes to a scrape as a whole, or the weight. The assumption is that said metrics aren't just some random collection but what is being retrieved from a target, and the purpose is to figure out what's taking up a scrape body. Cardinality isn't necessarily proportional to scrape size, which is why the weight is more useful, hence we call this command "check-weight".

Of course, it could go either way, and I respect your judgment.

**Comment by brian-brazil** [comment]
> is less about the cardinality of individual metrics and more about how that cardinality contributes to a scrape as a whole or the weight

You seem to be pulling terminology out of thin air here. You say that it's cardinality, and then pick an arbitrary term for it which has no bearing on its meaning, nor any history within this project. It does however have connotations in areas such as load balancing, that'd confuse users.

What I'd suggest is  `--analyze`, which also gives scope to expand in future.

> Cardinality isn't necessarily proportional to scrape size

Can you explain? The higher the cardinality, the bigger the scrape.

**Comment by beorn7** [comment]
I think the current treatment of histograms and summary is fine, i.e. counting each metric line in the text format rather than the whole histogram or summary as one. The former is how Prometheus currently ingests it, i.e. each bucket, quantile, and the sum and count creates a separate time series.

**Comment by beorn7** [comment]
About the naming discussion: I would also prefer to avoid introducing the term _weight_. That's more likely adding confusion than removing it.

We don't have a strict definition of _cardinality_ in the [glossary](https://prometheus.io/docs/introduction/glossary/). From what I can see in the docs, _cardinality_ is used with a fairly wide meaning (see for example the [tsdb-stats endpoint doc](https://prometheus.io/docs/prometheus/latest/querying/api/#tsdb-stats)). It makes sense to me to say "The metric family `api_requests_total` has a cardinality of 1234." if I want to say that there are 1234 time series with the name `api_requests_total`. I could also say "The label pair `{job="api_server"}` has a cardinality of 4321.", which would mean there are 4321 time series of any name with the label pair `{job="api_server"}`. And finally, it also makes sense to me to say "This scrape has a cardinality of 321.", which means to me that ingesting the scrape will add samples to 321 time series on the Prometheus server.

Since the output contains percentage and absolute counts, I don't think we have to include the percentage aspect in the naming discussion.

Finally, _weight_ could easily be understand as "weight in bytes", which it is not.

In different news, I like the proposal of using a more general term like `--analyze` so that we can add more information later without coming up with a new name. For consistency, perhaps this could be `--extended` as we use that already for `promtool tsdb analyze`, with a help text "Run extended analysis." (You can also see that `analyze` is already a command, so it would be confusing to use it as a flag somewhere else. Or we could do `promtool check metrics analyze`. Unfortunately, we are already somewhat inconsistent with the commands, sometimes starting with a verb, sometimes with a noun.)

**Comment by LeviHarrison** [comment]
I also like `check metrics --analyze`. I don't think we can do `check metrics analyze` because kingpin doesn't seem to support it (see https://github.com/prometheus/prometheus/pull/10045#issuecomment-1000854139).

**Comment by beorn7** [comment]
> 

> I don't think we can do `check metrics analyze` because kingpin doesn't seem to support it (see [#10045 (comment)](https://github.com/prometheus/prometheus/pull/10045#issuecomment-1000854139)).

Right, oh well… My vote is for `check metrics --extended` then, but I don't feel strongly.

**Comment by chenlujjj** [comment]
Thanks for the suggestions, I also prefer the `check metrics --extended` for consistency reason.

**Comment by chenlujjj** [comment]
Have updated the PR and fix CI :)

**Comment by beorn7** [comment]
Reminder: In a week (on 2022-01-12), I'll cut a new release (v2.33). I think this PR would be great to go into the release, so it would be good if the review could be concluded soon. It looks to me like we are almost there…

File: cmd/promtool/main.go [hunk 2]
```diff
@@ -645,9 +653,70 @@ func CheckMetrics() int {
...
 	}
 
+	if checkMetricsExtended {
+		stats, total, err := checkMetricsCount(&buf)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return failureExitCode
+		}
+		w := tabwriter.NewWriter(os.Stdout, 4, 4, 4, ' ', tabwriter.TabIndent)
+		fmt.Fprintf(w, "Metric\tCount\tPercentage\t\n")
```
Comment (LeviHarrison) [inline cmd/promtool/main.go:663]:
```suggestion
		fmt.Fprintf(w, "Metric\tCardinality\tPercentage\t\n")
```

**Review by LeviHarrison — COMMENTED** [review]
This is generally looking really good, mostly just wording changes. One of the most prominent ones is count -> cardinality, which I pointed out in a few places but should be changed everywhere.

**Review by JessicaGreben — APPROVED** [review]
thanks for all the work. looks good to me. once these final suggestions are implemented and @LeviHarrison gives the final approval, we can merge.

File: cmd/promtool/main.go [hunk 3]
```diff
@@ -645,9 +653,70 @@ func CheckMetrics() int {
...
+		fmt.Fprintf(w, "Total\t%d\t%.f%%\t\n", total, 100.)
+		w.Flush()
+	}
+
 	return successExitCode
 }
 
+type metricStat struct {
+	name       string
+	count      int
```
[the remark this answers is not in the document]
Comment (chenlujjj) [reply cmd/promtool/main.go:676]: missed that, fixed

**Comment by chenlujjj** [comment]
@LeviHarrison Thanks for the review work. All resolved now.

**Review by LeviHarrison — APPROVED** [review]

**Comment by LeviHarrison** [comment]
Thank you everyone for their work on this, and especially @chenlujjj.

**Comment by chenlujjj** [comment]
Thanks to all reviews and discussions, really learned a lot here. :)

**Pull request merged**